From 1c8721d96d5b5a2db72239ee67a2ec55bd447cfe Mon Sep 17 00:00:00 2001
From: Istio Automation <istio-testing-bot@google.com>
Date: Mon, 24 Jan 2022 08:54:07 -0800
Subject: [release-1.12] Recompute service account at endpoint shard when all
 endpoints are removed (#36940)

* Recompute service account at endpoint shard when all endpoints are removed.

* rn

Co-authored-by: Pengyuan Bian <bianpengyuan@google.com>
---
 pilot/pkg/xds/eds.go          | 12 ++++++++++--
 pilot/pkg/xds/eds_test.go     | 23 +++++++++++++++++++++++
 releasenotes/notes/36465.yaml |  9 +++++++++
 3 files changed, 42 insertions(+), 2 deletions(-)
 create mode 100644 releasenotes/notes/36465.yaml

diff --git a/pilot/pkg/xds/eds.go b/pilot/pkg/xds/eds.go
index 1a0dd0f642..e2551de96d 100644
--- a/pilot/pkg/xds/eds.go
+++ b/pilot/pkg/xds/eds.go
@@ -127,11 +127,11 @@ func (s *DiscoveryServer) EDSCacheUpdate(shard model.ShardKey, serviceName strin
 func (s *DiscoveryServer) edsCacheUpdate(shard model.ShardKey, hostname string, namespace string, istioEndpoints []*model.IstioEndpoint) bool {
 	if len(istioEndpoints) == 0 {
 		// Should delete the service EndpointShards when endpoints become zero to prevent memory leak,
-		// but we should not do not delete the keys from EndpointShardsByService map - that will trigger
+		// but we should not delete the keys from EndpointShardsByService map - that will trigger
 		// unnecessary full push which can become a real problem if a pod is in crashloop and thus endpoints
 		// flip flopping between 1 and 0.
 		s.deleteEndpointShards(shard, hostname, namespace)
-		log.Infof("Incremental push, service %s has no endpoints", hostname)
+		log.Infof("Incremental push, service %s at shard %v has no endpoints", hostname, shard)
 		return false
 	}
 
@@ -205,6 +205,14 @@ func (s *DiscoveryServer) deleteEndpointShards(shard model.ShardKey, serviceName
 		epShards := s.EndpointShardsByService[serviceName][namespace]
 		epShards.mutex.Lock()
 		delete(epShards.Shards, shard)
+		epShards.ServiceAccounts = sets.Set{}
+		for _, shard := range epShards.Shards {
+			for _, ep := range shard {
+				if ep.ServiceAccount != "" {
+					epShards.ServiceAccounts.Insert(ep.ServiceAccount)
+				}
+			}
+		}
 		// Clear the cache here to avoid race in cache writes (see edsCacheUpdate for details).
 		s.Cache.Clear(map[model.ConfigKey]struct{}{{
 			Kind:      gvk.ServiceEntry,
diff --git a/pilot/pkg/xds/eds_test.go b/pilot/pkg/xds/eds_test.go
index eaf910595c..71a68e0c6e 100644
--- a/pilot/pkg/xds/eds_test.go
+++ b/pilot/pkg/xds/eds_test.go
@@ -522,6 +522,29 @@ func TestUpdateServiceAccount(t *testing.T) {
 	}
 }
 
+func TestZeroEndpointShardSA(t *testing.T) {
+	cluster1Endppoints := []*model.IstioEndpoint{
+		{Address: "10.172.0.1", ServiceAccount: "sa1"},
+	}
+	s := new(xds.DiscoveryServer)
+	originalEndpointsShard := &xds.EndpointShards{
+		Shards: map[model.ShardKey][]*model.IstioEndpoint{
+			"c1": cluster1Endppoints,
+		},
+		ServiceAccounts: map[string]struct{}{
+			"sa1": {},
+		},
+	}
+	s.EndpointShardsByService = make(map[string]map[string]*xds.EndpointShards)
+	s.EndpointShardsByService["test"] = make(map[string]*xds.EndpointShards)
+	s.EndpointShardsByService["test"]["test"] = originalEndpointsShard
+	s.Cache = model.DisabledCache{}
+	s.EDSCacheUpdate("c1", "test", "test", []*model.IstioEndpoint{})
+	if len(s.EndpointShardsByService["test"]["test"].ServiceAccounts) != 0 {
+		t.Errorf("endpoint shard service accounts got %v want 0", len(s.EndpointShardsByService["test"]["test"].ServiceAccounts))
+	}
+}
+
 func fullPush(s *xds.FakeDiscoveryServer) {
 	s.Discovery.Push(&model.PushRequest{Full: true})
 }
diff --git a/releasenotes/notes/36465.yaml b/releasenotes/notes/36465.yaml
new file mode 100644
index 0000000000..8a0770da02
--- /dev/null
+++ b/releasenotes/notes/36465.yaml
@@ -0,0 +1,9 @@
+apiVersion: release-notes/v2
+kind: bug-fix
+area: traffic-management
+issue:
+  - https://github.com/istio/istio/issues/36456
+releaseNotes:
+  - |
+    **Fixed** an issue where scaling endpoint for a service from 0 to 1
+    might cause client side serivce account verification populated incorrectly.
-- 
2.35.3

